File: sdk-java/hfcx-sdk-core/src/main/java/eg/gov/healthflow/hfcx/sdk/core/validators/EgyptianIBANValidator.java
package eg.gov.healthflow.hfcx.sdk.core.validators;

import java.math.BigInteger;
import java.util.regex.Pattern;

/**
 * Validator for Egyptian IBANs.
 *
 * <p>Per CBE specification: {@code EG} + 2 check digits + 25 alphanumeric
 * characters (4 bank, 4 branch, 17 account) = 29 characters total. Must
 * pass the standard ISO 13616 mod-97 check.
 */
public final class EgyptianIBANValidator {

    private static final Pattern STRUCTURE = Pattern.compile("^EG\\d{2}[A-Z0-9]{25}$");
    private static final BigInteger NINETY_SEVEN = BigInteger.valueOf(97);

    private EgyptianIBANValidator() {}

    public static boolean isValid(String iban) {
        if (iban == null) {
            return false;
        }
        String cleaned = iban.replace(" ", "").toUpperCase();
        if (!STRUCTURE.matcher(cleaned).matches()) {
            return false;
        }
        // ISO 13616 mod-97: move first four chars to the end, replace each
        // letter with its 1-indexed position-in-the-alphabet plus 9
        // (A → 10, B → 11, ..., Z → 35), parse as a BigInteger, check mod 97 == 1.
        String reordered = cleaned.substring(4) + cleaned.substring(0, 4);
        StringBuilder numeric = new StringBuilder(reordered.length() * 2);
        for (int i = 0; i < reordered.length(); i++) {
            char c = reordered.charAt(i);
            if (Character.isDigit(c)) {
                numeric.append(c);
            } else {
                numeric.append(c - 'A' + 10);
            }
        }
        return new BigInteger(numeric.toString()).mod(NINETY_SEVEN).intValue() == 1;
    }
}

File: sdk-dotnet/src/HealthFlow.Hfcx.Sdk/Validators/EgyptianIbanValidator.cs
// Copyright (c) HealthFlow Medical HCX. Licensed under Apache 2.0.

using System;
using System.Numerics;
using System.Text;
using System.Text.RegularExpressions;

namespace HealthFlow.Hfcx.Sdk.Validators;

/// <summary>
/// Validator for Egyptian IBANs. Cross-SDK invariant with Java's
/// <c>EgyptianIBANValidator</c> and Python's <c>egyptian_iban</c>:
/// 29-char structural shape (<c>EG</c> + 2 check digits + 25
/// alphanumeric) plus the ISO 13616 mod-97 check.
/// </summary>
public static partial class EgyptianIbanValidator
{
    /// <summary>Returns <see langword="true"/> iff the IBAN is structurally valid and mod-97 = 1.</summary>
    public static bool IsValid(string? iban)
    {
        if (string.IsNullOrEmpty(iban))
        {
            return false;
        }

        var stripped = iban.Replace(" ", "", StringComparison.Ordinal).ToUpperInvariant();
        if (!Structure().IsMatch(stripped))
        {
            return false;
        }

        return Mod97(stripped) == 1;
    }

    private static int Mod97(string iban)
    {
        // Move the first four characters (country + check digits) to the end,
        // then convert each letter to two digits (A=10, B=11, ..., Z=35) and
        // compute mod 97 over the resulting numeric string.
        var rearranged = iban[4..] + iban[..4];
        var sb = new StringBuilder(rearranged.Length * 2);
        foreach (var c in rearranged)
        {
            if (c >= '0' && c <= '9')
            {
                sb.Append(c);
            }
            else
            {
                sb.Append((c - 'A' + 10).ToString(System.Globalization.CultureInfo.InvariantCulture));
            }
        }

        var big = BigInteger.Parse(sb.ToString(), System.Globalization.CultureInfo.InvariantCulture);
        return (int)(big % 97);
    }

    [GeneratedRegex(@"^EG\d{2}[A-Z0-9]{25}$")]
    private static partial Regex Structure();
}
